src/algorithms/xp.cpp

```diff
@@ -0,0 +1,612 @@
+            if (buffer == 'P') {
+                // We found the magic value!
+
+            } else {
+                // Put back both characters
+                in.unget();
+                in.unget();
+            }
+        } else {
+            // Put back the one character
+            in.unget();
+        }
+
+        try {
+            pos_map_iv.load(in);
+            sdsl::read_member(path_count, in);
+            pn_iv.load(in);
+            pn_csa.load(in);
+            pn_bv.load(in);
+            pn_bv_rank.load(in, &pn_bv);
+            pn_bv_select.load(in, &pn_bv);
+            pi_iv.load(in);
+
+            for (size_t i = 0; i < path_count; ++i) {
+                auto path = new XPPath;
+                // Load the path, giving it the file version and a
+                // rank-to-ID comversion function for format upgrade
+                // purposes.
+                path->load(in);
+                paths.push_back(path);
+            }
+        } catch (const XPFormatError &e) {
+            // Pass XGFormatErrors through
+            throw e;
+        } catch (const std::bad_alloc &e) {
+            // We get std::bad_alloc generally if we try to read arbitrary data as an xg index.
+            throw XPFormatError("XP input data not in XP version " + std::to_string(42) + " format (" + e.what() + ")");
+        } catch (const std::exception &e) {
+            // Other things will get re-thrown with a hint.
+            std::cerr << "error [xp]: Unexpected error parsing XP data. Is it in version " << "EASTER EGG"
+                      << " XP format?" << std::endl;
+            throw e;
+        }
+    }
+
+    bool XP::has_path(const std::string& path_name) const {
+        // find the name in the csa
+        std::string query = start_marker + path_name + end_marker;
+        auto occs = locate(pn_csa, query);
+        if (occs.size() > 1) {
+            std::cerr << "error [xp]: multiple hits for " << query << std::endl;
+            exit(1);
+        }
+        return !(occs.size() == 0);
+    }
+
+    bool XP::has_position(const std::string& path_name, size_t nuc_pos) const {
+        if (has_path(path_name)) {
+            const XPPath& xppath = get_path(path_name);
+            return !(xppath.offsets.size() < nuc_pos);
+        } else {
+            return false;
+        }
+    }
+
+    path_handle_t XP::get_path_handle(const std::string& path_name) const {
+        // find the name in the csa
+        std::string query = start_marker + path_name + end_marker;
+        auto occs = locate(pn_csa, query);
+        if (occs.size() > 1) {
+            std::cerr << "error [xp]: multiple hits for " << query << std::endl;
+            exit(1);
+        }
+        if(occs.size() == 0) {
+            // This path does not exist. Give back 0, which can never be a real path
+            // rank.
+            return as_path_handle(0);
+        }
+        return as_path_handle(pn_bv_rank(occs[0])+1); // step past '#'
```

Comment:
Like XG, you're adding 1 to the path handle here. This is why it's different than odgi. The difference is fine. We don't expect there to be stable handle definitions across different implementations.

Reply:
I see :)
Thanks Erik!